pubsub/google/cloud/pubsub_v1/publisher/batch.py

```diff
@@ -0,0 +1,231 @@
+            publisher client used to create this batch. Batch settings are
+            inferred from this.
+        topic (str): The topic. The format for this is
+            ``projects/{project}/topics/{topic}``.
+        settings (:class:`google.cloud.pubsub_v1.types.Batching`): The
+            settings for batch publishing. These should be considered
+            immutable once the batch has been opened.
+        autocommit (bool): Whether to autocommit the batch when the time
+            has elapsed. Defaults to True unless ``settings.max_latency`` is
+            inf.
+    """
+    def __init__(self, client, topic, settings, autocommit=True):
+        self._client = client
+
+        # Create a namespace that is owned by the client manager; this
+        # is necessary to be able to have these values be communicable between
+        # processes.
+        self._ = self.manager.Namespace()
+        self._.futures = self.manager.list()
+        self._.messages = self.manager.list()
+        self._.message_ids = self.manager.dict()
+        self._.settings = settings
+        self._.status = 'accepting messages'
+        self._.topic = topic
+
+        # This is purely internal tracking.
+        self._process = None
+
+        # Continually monitor the thread until it is time to commit the
+        # batch, or the batch is explicitly committed.
+        if autocommit and self._.settings.max_latency < float('inf'):
+            self._process = self._client.thread_class(target=self.monitor)
+            self._process.start()
+
+    @property
+    def client(self):
+        """Return the client that created this batch.
+
+        Returns:
+            :class:~`pubsub_v1.client.Client`: The client that created this
+                batch.
+        """
+        return self._client
+
+    @property
+    def manager(self):
+        """Return the client's manager.
+
+        Returns:
+            :class:`multiprocessing.Manager`: The manager responsible for
+                handling shared memory objects.
+        """
+        return self._client.manager
+
+    @property
+    def status(self):
+        """Return the status of this batch.
+
+        Returns:
+            str: The status of this batch. All statuses are human-readable,
+                all-lowercase strings.
+        """
+        return self._.status
+
+    def commit(self):
+        """Actually publish all of the messages on the active batch.
+
+        This moves the batch out from being the active batch to an in-flight
+        batch on the publisher, and then the batch is discarded upon
+        completion.
+        """
+        # Update the status.
+        self._.status = 'in-flight'
+
+        # Begin the request to publish these messages.
+        if len(self._.messages) == 0:
+            raise Exception('Empty queue')
+        response = self._client.api.publish(self._.topic, self._.messages)
+
+        # FIXME (lukesneeringer): Check for failures; retry.
```

Comment:
I need to go back and determine what errors there are. Also, retries need to be a thing (most likely that will be a decorator on `commit`.)